### DISCUSSION TITLE
Variables map reusability among different Expression instances

### DISCUSSION THREAD

**Discussion by oswaldobapvicjr** [ezylang/EvalEx#419] [Q&A] [answered]
Hello,
I have just tested EvalEx. You guys did quite a good job!
But I have a scenario where I need several independent (still sequential) `Expression` instances and I couldn't find a way to share the same hash map of variables among the Expression instances.

I am using the method `Expression.vithValues(...)`, which accepts a `Map<String, ?>` of preset values as a parameter, which, inside, will perform value conversions (to `EvaluationValue`).

I couldn't find a way to pass a hashmap of already converted `EvaluationValues`, so it seems that conversion will happen several times for the same variables when I create new `Expression` instances, which sounds like a performance waste for me.

Is there a way to avoid that?
Thank you very much,

**Comment by uklimaschewski** [comment, marked as answer]
If you pass an instance of an `EvaluationValue`, then no conversion will be done.
Since version 3.1.0, this is done in the `DefaultEvaluationValueConverter#convertObject()` method.

Another approach for you might be to create your own custom data accessor, replacing the default `MapBasedDataAccessor`.
See [Data Access](https://ezylang.github.io/EvalEx/customization/data_access.html) chapter in the documentation.

**Comment by tsuki-0101** [comment]
@oswaldobapvicjr can you let me know how did you approached this?

**Reply by oswaldobapvicjr** [reply]
I created my own, custom `DataAccessor` and implemented the Adapter design pattern, i.e, the accessor class containing a reference to my context map.
